atomicwriter/atomicwriter.go

```diff
@@ -0,0 +1,242 @@
+// Package atomicwriter provides utilities to perform atomic writes to a
+// file or set of files.
+package atomicwriter
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/moby/sys/sequential"
+)
+
+func validateDestination(fileName string) error {
+	if fileName == "" {
+		return errors.New("file name is empty")
+	}
+
+	// Deliberately using Lstat here to match the behavior of [os.Rename],
+	// which is used when completing the write and does not resolve symlinks.
+	//
+	// TODO(thaJeztah): decide whether we want to disallow symlinks or to follow them.
+	if fi, err := os.Lstat(fileName); err != nil {
+		if !os.IsNotExist(err) {
+			return fmt.Errorf("failed to stat output path: %w", err)
+		}
+	} else if err := validateFileMode(fi.Mode()); err != nil {
+		return err
+	}
+	if dir := filepath.Dir(fileName); dir != "" && dir != "." {
+		if _, err := os.Stat(dir); errors.Is(err, os.ErrNotExist) {
+			return fmt.Errorf("invalid file path: %w", err)
+		}
+	}
+	return nil
+}
+
+func validateFileMode(mode os.FileMode) error {
+	switch {
+	case mode.IsRegular():
+		return nil // Regular file
+	case mode&os.ModeDir != 0:
+		return errors.New("cannot write to a directory")
+	// TODO(thaJeztah): decide whether we want to disallow symlinks or to follow them.
+	// case mode&os.ModeSymlink != 0:
```

Comment:
Maybe we can get rid of this todo and just error out unless we are explicitly going to add a case to follow symlinks along with a test.

Reply:
It's a slightly tricky one; if we want this writer to be a drop-in for a regular `os.WriteFile`, then we should probably follow the symlink. We could resolve the symlink when creating the writer, then use that value as destination, but of course we may need to check where it's currently used (in case this would open up some symlink attack vector in moby?)

Erroring out would (in the meantime?) also be fine; currently it's kinda horrible, because if the destination is a symlink, the `os.Rename` just replaces the symlink with a file, which is not very expected I guess.

Reply:
fixed in https://github.com/moby/moby/pull/49741 - now producing an error (at least for now) for symlinked target